### DISCUSSION TITLE
Validator - FATAL error, SSL connection

### DISCUSSION THREAD

**Discussion by mtomorowicz** [FirelyTeam/firely-net-sdk#1833] [Q&A] [answered]
Hi guys!

We are working with the latest version(s) of your library ("Hl7.Fhir.R5" Version="3.4.0" & "Hl7.Fhir.Specification.R5" Version="3.4.0") and seeing bunch of FATALs as a result of validation process, see image below:

![image](https://user-images.githubusercontent.com/4983014/127990504-b9f5d7da-ad1f-4bee-b95d-7381b2d56f8b.png)

I was not able to find much information about this issue online, the exception message itself does not explain why, in cases where the system's node value is a generally available url (e.g. https://terminology.hl7.org/CodeSystem/v3-DataOperation), we are still experiencing SSL connection issues (and from our understanding those urls are supposed to be URNs ? we do have a separate code system definition file defined for this which seems to be ignored)... 

Here is a snippet of our XML with resource (Provenance) nodes resulting in two FATALs seen in the screenshot above:

```
<Provenance>
  <id value="prov-id"/>
  <target>
     <reference value="PackagedProductDefinition/omitted-on-purpose"/> 
  </target>
  <occurredDateTime value="2021-03-29T10:10:00Z"/>
  <recorded value="2021-03-30T10:10:00Z"/>
  <reason>
     <coding>
        <system value="https://spor.ema.europa.eu/v1/lists/100000152091"/>
        <code value="100000152447"/>
        <display value="_Redacted out_"/>
     </coding>
  </reason>
  <!-- what sort of change (edit, add) -->
  <activity>
     <coding>
        <system value="https://terminology.hl7.org/CodeSystem/v3-DataOperation"/>
        <code value="CREATE"/>
     </coding>
  </activity>
  <agent>
     <who>
        <reference value="PractitionerRole/applicant-person-id"/>
     </who>
  </agent>
</Provenance>
```
(note the https://spor.ema.europa.eu/v1/lists/100000152091 [requires auth] and https://terminology.hl7.org/CodeSystem/v3-DataOperation [generally available] pointers)

Would you guys be able to share some insights as to why we could be seeing those errors please?

Thanks

**Comment by marcovisserFurore** [comment, marked as answer]
Hi @mtomorowicz,

I tried to reproduce your validation error with the following unit test:

```c#
        public void Issue_1833()
        {
            var provenanceBundleXml = File.ReadAllText(Path.Combine("TestData", "validation", "provenanceBundle.xml"));

            var provenanceBundle = new FhirXmlParser().Parse<Bundle>(provenanceBundleXml);
            Assert.NotNull(provenanceBundle);

            var settings = new ValidationSettings()
            {
                ResourceResolver = ZipSource.CreateValidationSource(),
                GenerateSnapshot = true,
                EnableXsdValidation = true,
                ResolveExternalReferences = true
            };
            var validator = new Validator(settings);
            var report = validator.Validate(provenanceBundle);
            report.Issue.Should().NotContain(i => i.Details.Text.Contains("SSL"));
        }
```

And I think you are validating a Bundle here, so I put the Provenance resource in a bundle:

```xml
<?xml version="1.0" encoding="utf-8" ?>
<Bundle xmlns="http://hl7.org/fhir">
  <type value="collection"/>
  <entry>
    <fullUrl value="http://example.org/fhir/Organization/23"/>
    <resource>
      <Provenance xmlns="http://hl7.org/fhir">
        <id value="prov-id"/>
        <target>
          <reference value="PackagedProductDefinition/omitted-on-purpose"/>
        </target>
        <occurredDateTime value="2021-03-29T10:10:00Z"/>
        <recorded value="2021-03-30T10:10:00Z"/>
        <reason>
          <coding>
            <system value="https://spor.ema.europa.eu/v1/lists/100000152091"/>
            <code value="100000152447"/>
            <display value="_Redacted out_"/>
          </coding>
        </reason>
        <!-- what sort of change (edit, add) -->
        <activity>
          <coding>
            <system value="https://terminology.hl7.org/CodeSystem/v3-DataOperation"/>
            <code value="CREATE"/>
          </coding>
        </activity>
        <agent>
          <who>
            <reference value="PractitionerRole/applicant-person-id"/>
          </who>
        </agent>
      </Provenance>
    </resource>
  </entry>
</Bundle>
```
 
In my case (R5, version 3.4.0) this unit test is succeeding. I suspect you are using other resource resolvers as well, maybe that's creating the errors. In other words: what are you doing different from me?

**Reply by mtomorowicz** [reply]
Thanks for such a prompt response @marcovisserFurore ! 

Your test does pass indeed, but:

![image](https://user-images.githubusercontent.com/4983014/128128051-381cb7c4-2857-4c8b-893d-8b364a09c380.png)
 
Do we know why it is not capable of resolving a ref to standard Bundle at all ?

Thanks!

**Reply by marcovisserFurore** [reply]
Hmm, it should resolve the profile `Bundle`, because that is one of the standard FHIR resources. The line

```c#
ResourceResolver = ZipSource.CreateValidationSource(),
```
should take care of this. It will search for the `specification.zip` on your system and unzip this in a temporary folder and use this folder to resolve the standard FHIR profiles. Sometimes it helps to remove the temporary folders `%Temp%\FhirArtifactCache*` on your system. Or point to the `specification.zip` yourself:

```c#
ResourceResolver = new ZipSource(@"<path>\specification.zip", new DirectorySourceSettings { IncludeSubDirectories = true }),
```
where `<path>` is the full path to the specification.zip. 

The file `specification.zip` is shipped with the package `Hl7.Fhir.Specification.R5` and should be somewhere in your solution.

**Reply by mtomorowicz** [reply]
Excellent tip, that cache location mention @marcovisserFurore ! With cache nuked and refreshed all seems to be in order now...

Massive thanks for your time and information, appreciate it.

Side question - do you guys have a doc / wiki page somewhere describing the entire validation process, approaches, best practices etc? I was not able to find it online...

**Reply by marcovisserFurore** [reply]
We have some documentation about our Firely .NET SDK on our site: https://docs.fire.ly/projects/Firely-NET-SDK/start.html

I can also highly recommend the videos of Gino Canessa (Microsoft). He made some nice coding instruction videos with our SDK. There are also some sessions about validation: https://youtu.be/yMV5-fEHdMA.

**Reply by mtomorowicz** [reply]
Thanks, docs at https://docs.fire.ly/projects/Firely-NET-SDK/start.html do not have a dedicated validation section, think it would be super beneficial to future consumers of your awesome lib if such a doc existed... Thanks again !

**Reply by ewoutkramer** [reply]
We fully agree, and have added it as a work item: https://github.com/FirelyTeam/firely-net-sdk/issues/1841!
